### PR TITLE
[WIP] Slim v4 Docs

### PR DESCRIPTION
**Pull request by l0gicgate**
This is the initial PR for the new docs in preparation for the upcoming major release of Slim 4. Please let me know of any changes you would like done.

You can see a live version of the docs at [http://slim-website.lgse.com](http://slim-website.lgse.com)

**Key Changes to Docs App**:
- Refactored `_config.yml` to accomodate future major releases
- Added windows support to `Gemfile` for windows users to be able to build locally
- Updated `README` to add Windows instructions as well as Build instructions
- Added `slim_version` to `_config.yml` which depicts the current version of the docs and is used throughout varying components in the layout to automate url generation for default links on the landing page.
- Added `slim_docs_versions` to `_config.yml` which is a comma separated string listing all of the available versions of the docs. In future major releases simply add another to the list which will automatically add it as an option to the `version-selector` component
- Doc version selector has been moved from the landing page of each docs to the left navigation menu for easier access
- Added `package-lock.json` to `.gitignore`

**New Docs Added**
- Added new `Packaged Middleware` section reflecting all the new middleware added mentioned in the [CHANGELOG](https://github.com/slimphp/Slim/blob/4.x/CHANGELOG.md)
- Added most of all new changes to the `Upgrade Guide`
- Updated deployment section to reflect changes to error handling
- Updated dependency injection section to reflect removal of Slim Container

**Fixes**
- Fix for #292: Added current version browsed in header
- Fix for #299: Fixed rewrite rule as per comment in deployment section

**Progress**
- [x] Draft new docs
- [ ] Input from contributors
- [ ] Code Review

### CODE REVIEW

**Comment by llvdl** [comment]
The installation page (`/docs/v4/start/installation.html`) should be updated:

> PHP 5.5 or newer

should be "Slim 4 requires PHP 7.0+" (match what is stated on the Upgrade Guide) page

From:
> `composer require slim/slim "^3.0"`

To:
    `composer require slim/slim "^4.0"`


The upgrade page ('/docs/v4/start/upgrade.html) should state that Slim 4 requires PHP 7.0+, not 7.1+.

**Comment by llvdl** [comment]
I would move the version selector to the top right, or make it possible to switch the version in the top right as well.

**Comment by l0gicgate** [comment]
Thank you @llvdl fixed both of your observations. As for the version selector I don't think it's necessary to have it available at all times. how often will you be switching from version to version? If others want this change I can do it though it's not a problem.

**Comment by llvdl** [comment]
> Thank you @llvdl fixed both of your observations. As for the version selector I don't think it's necessary to have it available at all times. how often will you be switching from version to version? If others want this change I can do it though it's not a problem.

Thanks for the great work on the documentation.

Regarding the language switch, it is just a suggestion, I am ok with the current solution. I like the new option to be able to switch versions from any page. When you open a search engine result, you can end up on any page, so it is handy to be able to switch easily. I just expected to be able to switch the version using the green button on the top right, instead of on the left in the navigation pane.

**Comment by tflight** [comment]
Thanks for your work on this, @l0gicgate, I really appreciate it! With regards to the fix for #292, that is certainly an improvement. I wonder if there might be a way to get the current major stable version (3 today) and if the version of the docs you are looking at isn't that version, color the box/button in red to make it more obvious you're not looking at the current version of the docs?

**Comment by l0gicgate** [comment]
@tflight definitely, I would probably pick some sort of yellow or deep orange to highlight if it is not the current version. I will add that soon! Thanks

File: docs/v4/start/upgrade.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+## New Middleware Approach
+In Slim 4 we wanted to give more flexibility to the developers by decoupling some of Slim's App core functionality and implementing it as middleware. This gives you the ability to swap in custom implementations of the core components.
+
+## Middleware Execution
+Middleware execution has not changed and is still Last In First Out (LIFO) like in Slim 3.
+
+## New Routing Middleware
+The routing has been implemented as middleware. We are still using [FastRoute](https://github.com/nikic/FastRoute) as the default router.
+We provide an instantiation of FastRoute via the App::getRouter() method out of the box. You will need to instantiate and add the RoutingMiddleware in order for the app to work.
+If you were using `determineRouteBeforeAppMiddleware`, you need to add the `Middleware\RoutingMiddleware`e middleware to your application just before your call run() to maintain the previous behaviour.
```
Comment (sander-bol) [inline docs/v4/start/upgrade.md]: Stray 'e' after `Middleware\RoutingMiddleware`.

**Comment by sander-bol** [comment]
You're probably aware of it, but the tutorial ( http://slim-website.lgse.com/docs/v4/tutorial/first-app.html ) does not seem to be fully V4 yet, especially the examples around application construction... there's still config there for Framework features that have been migrated to Middleware.

**Comment by akrabat** [comment]
Can you create an issue for updating the tutorial please?

**Comment by l0gicgate** [comment]
The tutorial is still in a limbo state, I'm waiting until we have everything buttoned up to rewrite it.

**Pull request merged**